fix: Error UI improvements for nested lists/objects

kwyoung11 opened the pull request:
<!--
Thanks for submitting a pull request!

Please make sure you've read and understood our contributing guidelines.

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx", where #xxxx is the issue number.

Please provide enough information so that others can review your pull request.
The first two fields are mandatory:
-->
fixes #2530
fixes #2791

**Summary**
This PR introduces a mechanism for building a tree representation of list and object widgets. This tree representation allows to determine the ancestors of a field that is nested inside list or object widgets, in order to highlight them in the UI on a validation error.

### Notes
- adds an `entryTreeMap` field to the state of the entryDraft.js reducer. 
- adds ADD_TO_ENTRY_TREE_MAP, REMOVE_FROM_ENTRY_TREE_MAP and SWAP_NODES_IN_ENTRY_TREE_MAP  actions in `entries.ts`
- these actions are called from within EditorControl and ListControl to build the tree
- utilized https://github.com/lukasbuenger/immutable-treeutils package to simplify building and operating on the tree.

Please let me know any feedback on the approach. 

There were also **other potential enhancements** mentioned in #2530, mainly: 

> One helpful solution would be to provide a link next to the message that takes you straight to the empty field. This would save a lot of time searching, but it would only really work for situations where a single field is empty.

and

> At the very least, it would be very nice if the dialog indicated how many fields are empty ("Oops, you are missing 3 required fields") or even what they are called ("Oops, you are missing 3 required fields: Post Title, Post Body, Post Categories").

that I think could be rolled into this PR?

<!--
Explain the **motivation** for making this change.
What existing problem does the pull request solve?
-->
**Motivation**
If you receive a validation error on a field that is nested inside of a list, typed list or object widget, you cannot see where the error is if the given widget is collapsed, because the UI does not give any indication. This PR adds error coloring to labels and borders of these widgets on validation error of any nested field.

**Test plan**
I've added a unit test for the new redux action.
I've added some basic Cypress integration tests, but need to add a few more.

### TODO
- more definition around other enhancements mentioned in #2530

<!--
Demonstrate the code is solid.
Example: The exact commands you ran and their output, screenshots / videos if the pull request changes UI.
-->

**A picture of a cute animal (not mandatory but encouraged)**

kwyoung11 commented:
**Screenshots of Change in UI before and after PR below:**

### **Before PR:**

**Fully Open List with Errors**: ![Screen Shot 2020-05-09 at 1 34 45 AM](https://user-images.githubusercontent.com/1297966/81466784-ce933880-9199-11ea-96fc-16c9a1f7e0dc.png)
**Partially Collapsed List with Errors**: ![Screen Shot 2020-05-09 at 1 35 10 AM](https://user-images.githubusercontent.com/1297966/81466794-e4a0f900-9199-11ea-942c-43367cba3845.png)
**Fully Collapsed List with Errors**: ![Screen Shot 2020-05-09 at 1 35 24 AM](https://user-images.githubusercontent.com/1297966/81466796-eb2f7080-9199-11ea-8638-65e2531a9739.png)



### **After PR:**
**Fully Open List with Errors**: ![Screen Shot 2020-05-09 at 1 27 51 AM](https://user-images.githubusercontent.com/1297966/81466693-34cb8b80-9199-11ea-9340-ca141c286a9f.png)
**Partially Collapsed List with Errors**: ![Screen Shot 2020-05-09 at 1 28 15 AM](https://user-images.githubusercontent.com/1297966/81466706-414fe400-9199-11ea-9d48-d548a7f3c635.png)
**Fully Collapsed List with Errors**: 
![Screen Shot 2020-05-09 at 1 28 45 AM](https://user-images.githubusercontent.com/1297966/81466712-4ca30f80-9199-11ea-9765-52cc25980364.png)

barthc commented:
I don't think it would work for `object` widget just `list` or should we care since the `object`  widget doesn't cascade.

erezrokah commented:
Thanks for this @kwyoung11 and for adding tests!
Just to make sure I understand, the main challenge here is not having the relevant ids to check `fieldsErrors` for errors in the list widget, hence the need for the new data structure?

I need to think about this since we are already passing down a lot of information.
Will do some digging and update here.

kwyoung11 commented:
@erezrokah correct. `fieldsErrors` just stores the ids of regular fields (not list or object widgets). The challenge is that you could have a list (list C) within a list (list B) within a list (list A), with list C containing a field with a validation error. So the question is how can you know, from list A, if there is a validation error in any descendant field. There didn't seem to be a straight forward way to do that to me from, for example, this method: https://github.com/netlify/netlify-cms/blob/master/packages/netlify-cms-core/src/components/Editor/EditorControlPane/Widget.js#L99-L118, which is where any validation errors are passed back to `fieldsErrors` via the `onValidate` prop.

Representing it as a tree in a separate data structure seemed like the simplest/cleanest solution to me. If you think of a better way to fix the issue though then that totally works too. I can also dig some more into it if you think we need to flesh it out some more.

@barthc This should work for object widgets, I added tests for that. I also manually tested on the object widgets in the kitchenSink collection.

barthc commented on packages/netlify-cms-core/src/reducers/entryDraft.js:
```diff
@@ -156,6 +180,60 @@ const entryDraftReducer = (state = Map(), action) => {
+          );
+          state.updateIn(['entryTreeMap', ...keyPathToParent, 'childNodes', newIndex], node =>
+            node.set('id', newId),
+          );
```
Can these be combined into a single operation?

barthc commented:
> fieldsErrors just stores the ids of regular fields (not list or object widgets). 

The `fieldsErrors` object can be modified to hold the parent id as key to descendant field.

kwyoung11 commented:
@barthc got it, I'll try to make that modification and push it up here soon.

kwyoung11 commented:
@erezrokah @barthc I removed entryTreeMap from redux state and added a `parentIds` prop to EditorControl that is passed down through Widget -> ListControl -> ObjectControl that keeps track of the parents of a field in an Object or List widget. It is stored inside of `fieldsErrors` in the corresponding error object. Thoughts?

erezrokah commented on packages/netlify-cms-core/src/components/Editor/EditorControlPane/EditorControl.js:
```diff
@@ -115,13 +115,31 @@ class EditorControl extends React.Component {
+  isAncestorOfFieldError = () => {
+    const { fieldsErrors } = this.props;
+
+    if (this.isListOrObject && fieldsErrors.size > 0) {
```
We should avoid hardcoding widget names in the core. Is `isListOrObject` required?

kwyoung11 replied:
I just put it there as a potential optimization so that it didn't need to check the parentIds for every single field, but it's probably not necessary.

erezrokah commented on packages/netlify-cms-widget-list/src/ListControl.js:
```diff
@@ -398,7 +399,10 @@ export default class ListControl extends React.Component {
     // Reset item to ensure updated state
     const updatedKeys = keys.map((key, keyIndex) => {
-      if (keyIndex === oldIndex || keyIndex === newIndex) {
+      if (keyIndex === oldIndex) {
```
We can merged this conditions into a single `||` condition

erezrokah commented on packages/netlify-cms-widget-object/src/ObjectControl.js:
```diff
@@ -79,13 +80,16 @@ export default class ObjectControl extends React.Component {
     }
     const fieldName = field.get('name');
     const fieldValue = value && Map.isMap(value) ? value.get(fieldName) : value;
+    const newParentIds = [...parentIds, forID, listControlId].filter(id => id !== null);
```
How about doing this in `ListControl` and only pass in `parentIds`
